sycl/doc/CompileTimeProperties.md

```diff
@@ -0,0 +1,813 @@
+    "sycl-foo",       // Name of second property
+    nullptr,          // First property has no parameter
+    32                // Value of second property
+    )]]
+#endif
+  ;
+};
+
+} // namespace sycl
+```
+
+As the name implies, this C++ attribute is only used to decorate a member
+variable of a class type that is as SYCL "special class" (i.e. a class that is
+decorated with `__attribute__((sycl_special_class))`).  The device compiler
+front-end ignores the attribute when it is used in any other syntactic
+position.
+
+The device compiler front-end uses this attribute only when the class type
+containing the decorated member variable is the type of a kernel argument,
+and it silently ignores the attribute when the class is used in any other way.
+
+When the front-end creates a kernel argument from a SYCL "special class", it
+passes each member variable of the class as a separate kernel argument.  If the
+member variable is decorated with
+`[[__sycl_detail__::add_ir_kernel_parameter_attributes()]]`, the front-end adds
+one LLVM IR attribute to the kernel function's parameter for each property in
+the list.  For example, this can be done by calling
+[`Function::addParamAttrs(unsigned ArgNo, const AttrBuilder &)`][7].  As
+before, the IR attributes are added as strings, so the front-end must convert
+the property value to a string if it is not already a string.
+
+[7]: <https://llvm.org/doxygen/classllvm_1_1Function.html#a092beb46ecce99e6b39628ee92ccd95a>
+
+
+## Properties on kernel functions
+
+Compile-time properties can also be used to decorate kernel functions as with
+the [SYCL\_EXT\_ONEAPI\_KERNEL\_PROPERTIES][8] extension.  There are two ways
+the application can specify these properties.  The first is by passing a
+`property_list` parameter to the function that submits the kernel:
+
+[8]: <https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/KernelProperties/KernelProperties.asciidoc>
+
+```
+namespace sycl {
+
+class handler {
+  template <typename KernelName, typename KernelType, typename PropertyListT>
+  void single_task(PropertyListT properties, const KernelType &kernelFunc);
+};
+
+// namespace sycl
+```
+
+For example:
+
+```
+using sycl;
+using sycl::ext::oneapi;
+
+void foo(handler &cgh) {
+  cgh.single_task(
+    property_list{sub_group_size_v<32>, device_has_v<aspect::fp16>},
+    [=] {/* ... */});
+}
+```
+
+The second way an application can specify kernel properties is by adding a
+`properties` member variable to a named kernel function object:
+
+```
+using sycl;
+using sycl::ext::oneapi;
+
+class MyKernel {
+ public:
+  void operator()() {/* ... */}
+
+  static constexpr auto properties =
+    property_list{sub_group_size_v<32>, device_has_v<aspect::fp16>};
+};
+
+void foo(handler &cgh) {
+  MyKernel k;
+  cgh.single_task(k);
+}
+```
+
+Internally, the headers lower both cases to a wrapper class that expresses the
+properties as an internal C++ attribute, and the `operator()` of this class
```

Comment:
Please name the attribute here.

Reply:
Done in 403e6c964c4539197fc186aee9a9ce2ff271ebe7